### ISSUE TITLE
Tracking bug: gcc 7.2 fallout

### ISSUE THREAD

**Issue by chneukirchen** [voidlinux/void-packages#8842]
Link issues here where gcc 7.2 breaks things.

Issues found so far:

- `std::to_string` is not defined because `_GLIBCXX11_USE_C99_STDIO` is now unset in /usr/include/c++/7.2/x86_64-unknown-linux-gnu/bits/c++config.h

**Comment by pullmoll** [comment]
Looking at `libstdc++v3/configure` we could predefine `glibcxx_cv_c99_stdio_cxx11=yes` to have that macro defined again.

**Comment by chneukirchen** [comment]
I have a better fix in the works, hold on.

**Comment by chneukirchen** [comment]
5ca276d45e

**Comment by Dadie** [comment]
Currently I'm not able to compile C++17 Code using clang with libstdc++ as gcc and libstdc++ was updated was but clang (and libc++) wasn't. Dunno if this count as "fallout" but it got me for a second and made my CI-Server sad :'(

**Comment by chneukirchen** [comment]
Please show an example of code that breaks.

libstdc++ abi should not have changed.

**Comment by Dadie** [comment]
Here is an [example](https://gist.github.com/Dadie/fca728143f4f2d8811c8f81afb3cf93e) with outputs.

**Comment by chneukirchen** [comment]
Will be fixed with clang 5, compare https://bugs.archlinux.org/task/54276

**Comment by pullmoll** [comment]
When bootstrapping with gcc-7.2.0 binutils fails to link because of relocation R_X86_64_32 symbols in libstdc++.a
```
/usr/bin/ld: /usr/lib/gcc/x86_64-unknown-linux-gnu/7.2.0/../../../../lib/libstdc++.a(compatibility.o): relocation R_X86_64_32 against hidden symbol `_ZN4_VTVISt15basic_streambufIcSt11char_traitsIcEEE12__vtable_mapE' can not be used when making a shared object
/usr/bin/ld: final link failed: Nonrepresentable section on output
```
I read somewhere that a compiler option `-fcmode=large` should produce R_X86_64_64 symbols and perhaps we need to inject this flag into libstd++. A page describing the cmode flags is [here](https://www.quora.com/What-is-difference-between-tiny-small-medium-compact-large-and-huge-memory-model-in-C?share=1)

**Comment by chneukirchen** [comment]
Disabling gold during bootstrap is ok I guess... does a rebuild then work with gold?

**Comment by pullmoll** [comment]
> Disabling gold during bootstrap is ok I guess... does a rebuild then work with gold?

No, the error is the same when you try to build `binutils` later on.
This is why I think we're missing some configure option for how `libstdc++.a` is built.

**Comment by chneukirchen** [comment]
Working on a fix.

**Comment by chneukirchen** [comment]
Ok, I think the problem is how we now pass `-fno-PIE` when `nopie=1`.

I'm currently trying to remove the flag, and let gcc's configure script figure out it needs it (this seems to work so far, build is still running.)

**Comment by chneukirchen** [comment]
I think the fix is http://sprunge.us/IfaR

Please review, then I can commit it as _4.

**Comment by chneukirchen** [comment]
It fixes the binutils build for one.

**Comment by pullmoll** [comment]
This patch makes very much sense. Letting configure find the required flags is certainly better than patching around the configure scripts.

**Comment by Yann-Kaelig** [comment]
Sry, but this patch doesn't fix my bootstrap issue.

**Comment by pullmoll** [comment]
It does here: After building gcc-7.2.0_4 I can build binutils with `--enable-gold` in place.
Do you have your host gcc updated as well?

**Comment by Yann-Kaelig** [comment]
Host updated, bootstrap from source built.
Thx @chneukirchen and @pullmoll.

**Comment by IronBug** [comment]
Hello, people. I've got a serious breaking bug in Void-musl.

I'm not sure this is the same issue, but in musl build I have totally broken the whole GCC toolchain and binutils. Looks like it happened wit the last update of GCC.
I got errors the compiler could not find cc1 and 
>error: -fuse-linker-plugin, but liblto_plugin.so not found, although liblto_plugin.so is in the same directory
I added *liblto_plugin.so* to *LD_LIBRARY_PATH* and */usr/lib/gcc/x86_64-linux-musl/7.2* to *PATH*. But still get errors with linker:
>ld: cannot find crtbegin.o: No such file or directory
>ld: cannot find -lgcc
So the whole toolchain is cranked. Before the last update it was allright, I use GCC every day.
